Import necessary libraries

In [ ]:
import pandas as pd
import numpy as np
import math
import re
import matplotlib.pyplot as plt
from google.colab import drive

Connect to google drive to access input files

In [ ]:
drive.mount('/content/drive')

Mounted at /content/drive


Access the dataset file 'combined_data_1.txt' file

In [ ]:
data = pd.read_csv('/content/drive/My Drive/240 project/combined_data_1.txt', header = None, names = ['Cust_Id', 'Rating'], usecols = [0,1])
data['Rating'] = data['Rating'].astype(float)
print('Data shape: {}'.format(data.shape))
print(data.iloc[::5000000, :])

Data shape: (24058263, 2)
          Cust_Id  Rating
0              1:     NaN
5000000   2560324     4.0
10000000  2271935     2.0
15000000  1921803     2.0
20000000  1933327     3.0


In [ ]:
merge_dataset_nan = pd.DataFrame(pd.isnull(data.Rating))
merge_dataset_nan = merge_dataset_nan[merge_dataset_nan['Rating'] == True]
merge_dataset_nan = merge_dataset_nan.reset_index()

movie_np = []
movie_id = 1

for i,j in zip(merge_dataset_nan['index'][1:],merge_dataset_nan['index'][:-1]):
    # numpy approach
    temp = np.full((1,i-j-1), movie_id)
    movie_np = np.append(movie_np, temp)
    movie_id += 1

# Account for last record and corresponding length
# numpy approach
last_record = np.full((1,len(data) - merge_dataset_nan.iloc[-1, 0] - 1),movie_id)
movie_np = np.append(movie_np, last_record)

print('Movie numpy: {}'.format(movie_np))
print('Length: {}'.format(len(movie_np)))

Movie numpy: [1.000e+00 1.000e+00 1.000e+00 ... 4.499e+03 4.499e+03 4.499e+03]
Length: 24053764


In [ ]:
data.to_csv('/content/drive/My Drive/240 project/data.csv', index=False)


In [ ]:

movie_count = data.isnull().sum()[1]

cust_count = data['Cust_Id'].nunique() - movie_count

rating_count = data['Cust_Id'].count() - movie_count


print('Total pool: {:,} Movies, {:,} customers, {:,} ratings given'.format(movie_count, cust_count, rating_count))

Total pool: 4,499 Movies, 470,758 customers, 24,053,764 ratings given


In [ ]:
data = data[pd.notnull(data['Rating'])]
#print(len(merge_dataset))
data['Movie_Id'] = movie_np.astype(int)
data['Cust_Id'] = data['Cust_Id'].astype(int)
print(data.iloc[::5000000, :])

          Cust_Id  Rating  Movie_Id
1         1488844     3.0         1
5000996    501954     2.0       996
10001962   404654     5.0      1962
15002876   886608     2.0      2876
20003825  1193835     2.0      3825


In [ ]:
path = '/content/drive/My Drive/240 project/data.csv'

with open(path, 'w', encoding = 'utf-8-sig') as f:
  data.to_csv(f,index=False)

In [ ]:
import pandas as pd
import numpy as np
import math
import re
import matplotlib.pyplot as plt
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
new_data = pd.read_csv('/content/drive/My Drive/240 project/data.csv')

Create a dataframe from the combined_data1 file

In [ ]:
new_data.head(10)

Cust_Id  Rating  Movie_Id
0  1488844     3.0         1
1   822109     5.0         1
2   885013     4.0         1
3    30878     4.0         1
4   823519     3.0         1
5   893988     3.0         1
6   124105     4.0         1
7  1248029     3.0         1
8  1842128     4.0         1
9  2238063     3.0         1

In [ ]:
new_data = new_data[new_data['Rating'].notna()]

Remove duplicates

In [ ]:
new_data = new_data.drop_duplicates(['Cust_Id','Movie_Id'])

Keep only movies that have the rating more than 3

In [ ]:
new_data = new_data[new_data['Rating'] >=  3.0]  

In [ ]:
print("Total Data:")
print("Total number of movie ratings = "+str(new_data.shape[0]))
print("Number of unique users = "+str(len(np.unique(new_data["Cust_Id"]))))
print("Number of unique movies = "+str(len(np.unique(new_data["Movie_Id"]))))

Total Data:
Total number of movie ratings = 20496505
Number of unique users = 467134
Number of unique movies = 4499


Access the movie names from 'movie_titles.csv' file and combine both the datasets

In [ ]:
n=3
merge_dataset_title = pd.read_csv('/content/drive/My Drive/240 project/movie_titles.csv', header = None,encoding='ISO-8859-1',usecols=range(n),
                 lineterminator='\n')
merge_dataset_title.columns = ['movie_id', 'year', 'name']
merge_dataset_title.head(10)

movie_id    year                          name
0         1  2003.0               Dinosaur Planet
1         2  2004.0    Isle of Man TT 2004 Review
2         3  1997.0                     Character
3         4  1994.0  Paula Abdul's Get Up & Dance
4         5  2004.0      The Rise and Fall of ECW
5         6  1997.0                          Sick
6         7  1992.0                         8 Man
7         8  2004.0    What the #$*! Do We Know!?
8         9  1991.0      Class of Nuke 'Em High 2
9        10  2001.0                       Fighter

In [ ]:
df = pd.merge(new_data, merge_dataset_title[['movie_id', 'name']], left_on='Movie_Id', right_on='movie_id')
df.head()

Cust_Id  Rating  Movie_Id  movie_id             name
0  1488844     3.0         1         1  Dinosaur Planet
1   822109     5.0         1         1  Dinosaur Planet
2   885013     4.0         1         1  Dinosaur Planet
3    30878     4.0         1         1  Dinosaur Planet
4   823519     3.0         1         1  Dinosaur Planet

Drop duplicates

In [ ]:
df=df.drop(['Movie_Id', 'movie_id'], axis=1)

In [ ]:
path = '/content/drive/My Drive/240 project/merged.csv'

with open(path, 'w', encoding = 'utf-8-sig') as f:
  df.to_csv(f,index=False)

In [ ]:
import pandas as pd
import numpy as np
import math
import re
import matplotlib.pyplot as plt
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
final = pd.read_csv('/content/drive/My Drive/240 project/merged.csv')

In [ ]:
final.head()

Cust_Id  Rating             name
0  1488844     3.0  Dinosaur Planet
1   822109     5.0  Dinosaur Planet
2   885013     4.0  Dinosaur Planet
3    30878     4.0  Dinosaur Planet
4   823519     3.0  Dinosaur Planet

Remove all duplicate records i.e., if same customer gave rating to any movie more than once remove them

In [ ]:
final = final.drop_duplicates(['Cust_Id','name'])

In [ ]:
len(final)

20477176

Total records after cleaning are 20477176

Create pivot table from the dataframe

In [ ]:
df_pivot = final.pivot(index='Cust_Id', columns='name', values='Rating').fillna(0)

convert pivot table values to an integer and replace the ratings with 1

In [ ]:
df_pivot = df_pivot.astype('int64')

In [ ]:
def encode_ratings(x):
    if x<=0:
        return 0
    if x>=1:
        return 1

df_pivot = df_pivot.applymap(encode_ratings)

Pivot table generated as below:

In [ ]:
df_pivot.head()

name     10  10 Things I Hate About You  \
Cust_Id                                   
6         0                           0   
7         0                           0   
79        0                           0   
97        0                           0   
134       0                           1   

name     101 Dalmatians II: Patch's London Adventure  11:14  13 Ghosts  1984  \
Cust_Id                                                                        
6                                                  0      0          0     1   
7                                                  0      0          0     0   
79                                                 1      0          0     0   
97                                                 0      0          0     0   
134                                                0      0          0     0   

name     2 Fast 2 Furious  200 Cigarettes  2010: The Year We Make Contact  \
Cust_Id                                                                     
6                       0               0                               0   
7                       1               0                               0   
79                      0               0                               0   
97                      0               0                               0   
134                     1               0                               0   

name     24 Hour Party People  ...  Xena: Warrior Princess: Season 3  \
Cust_Id                        ...                                     
6                           0  ...                                 0   
7                           0  ...                                 0   
79                          0  ...                                 0   
97                          0  ...                                 0   
134                         0  ...                                 0   

name     Xena: Warrior Princess: Series Finale  Y Tu Mama Tambien  \
Cust_Id                                                             
6                                            0                  0   
7                                            0                  0   
79                                           0                  0   
97                                           1                  0   
134                                          0                  0   

name     Yellow Submarine  Yi Yi  Yojimbo  Young Black Stallion  Youngblood  \
Cust_Id                                                                       
6                       0      0        0                     0           0   
7                       0      0        0                     0           0   
79                      0      0        0                     0           0   
97                      0      0        0                     0           0   
134                     0      0        0                     0           0   

name     Yu-Gi-Oh!: The Movie  Zorro  
Cust_Id                               
6                           0      0  
7                           0      0  
79                          1      0  
97                          0      0  
134                         0      0  

[5 rows x 1340 columns]

In [ ]:
path = '/content/drive/My Drive/240 project/pivot.csv'

with open(path, 'w', encoding = 'utf-8-sig') as f:
  df_pivot.to_csv(f)

In [ ]:
df_pivot = pd.read_csv('/content/drive/My Drive/240 project/pivot_3.csv',index_col=0)

In [ ]:
len(df_pivot)

140361

Pivot table has total 140361 records

In [ ]:
df_pivot.head()

10  10 Things I Hate About You  \
Cust_Id                                   
6         0                           0   
7         0                           0   
79        0                           0   
97        0                           0   
134       0                           1   

         101 Dalmatians II: Patch's London Adventure  11:14  13 Ghosts  1984  \
Cust_Id                                                                        
6                                                  0      0          0     1   
7                                                  0      0          0     0   
79                                                 1      0          0     0   
97                                                 0      0          0     0   
134                                                0      0          0     0   

         2 Fast 2 Furious  200 Cigarettes  2010: The Year We Make Contact  \
Cust_Id                                                                     
6                       0               0                               0   
7                       1               0                               0   
79                      0               0                               0   
97                      0               0                               0   
134                     1               0                               0   

         24 Hour Party People  ...  Xena: Warrior Princess: Season 3  \
Cust_Id                        ...                                     
6                           0  ...                                 0   
7                           0  ...                                 0   
79                          0  ...                                 0   
97                          0  ...                                 0   
134                         0  ...                                 0   

         Xena: Warrior Princess: Series Finale  Y Tu Mama Tambien  \
Cust_Id                                                             
6                                            0                  0   
7                                            0                  0   
79                                           0                  0   
97                                           1                  0   
134                                          0                  0   

         Yellow Submarine  Yi Yi  Yojimbo  Young Black Stallion  Youngblood  \
Cust_Id                                                                       
6                       0      0        0                     0           0   
7                       0      0        0                     0           0   
79                      0      0        0                     0           0   
97                      0      0        0                     0           0   
134                     0      0        0                     0           0   

         Yu-Gi-Oh!: The Movie  Zorro  
Cust_Id                               
6                           0      0  
7                           0      0  
79                          1      0  
97                          0      0  
134                         0      0  

[5 rows x 1340 columns]

In [ ]:
# Import library
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

Generate frequent itemsets with minimum support of 30%

In [ ]:
from mlxtend.frequent_patterns import apriori

frequent_itemset = apriori(df_pivot, min_support=0.30, use_colnames=True)

In [ ]:
len(frequent_itemset)

939

Total 939 frequent itemsets are generated

In [ ]:
frequent_itemset.tail()

support                                           itemsets
934  0.306666  (The Last Samurai, The Silence of the Lambs, P...
935  0.300589  (The Sixth Sense, Lord of the Rings: The Fello...
936  0.308547  (The Sixth Sense, The Silence of the Lambs, Lo...
937  0.305534  (The Sixth Sense, Lord of the Rings: The Fello...
938  0.303831  (The Sixth Sense, The Silence of the Lambs, Lo...

In [ ]:
path = '/content/drive/My Drive/240 project/freqitemset30.csv'

with open(path, 'w', encoding = 'utf-8-sig') as f:
  frequent_itemset.to_csv(f)

In [ ]:
frequent_itemset = pd.read_csv('/content/drive/My Drive/240 project/freqitemset30.csv',index_col=0)

Itemsets and their support are shown below:

In [ ]:
frequent_itemset.head()

support            itemsets
0  0.548799    (50 First Dates)
1  0.547146  (A Beautiful Mind)
2  0.326886     (About Schmidt)
3  0.351102       (About a Boy)
4  0.390201  (Along Came Polly)

Generate association rules using lift with threshold of 1

In [ ]:
from mlxtend.frequent_patterns import association_rules

rules = association_rules(frequent_itemset, metric="lift", min_threshold=1)

In [ ]:
len(rules)

4266

Total 4266 rules are generated

In [ ]:
rules.head()

antecedents         consequents  antecedent support  \
0  (A Beautiful Mind)    (50 First Dates)            0.547146   
1    (50 First Dates)  (A Beautiful Mind)            0.548799   
2    (50 First Dates)  (Along Came Polly)            0.548799   
3  (Along Came Polly)    (50 First Dates)            0.390201   
4        (Braveheart)    (50 First Dates)            0.619930   

   consequent support   support  confidence      lift  leverage  conviction  
0            0.548799  0.317268    0.579859  1.056596  0.016994    1.073927  
1            0.547146  0.317268    0.578112  1.056596  0.016994    1.073399  
2            0.390201  0.314667    0.573374  1.469433  0.100525    1.429353  
3            0.548799  0.314667    0.806423  1.469433  0.100525    2.330864  
4            0.548799  0.365600    0.589744  1.074608  0.025383    1.099804

In [ ]:
df_res = rules.sort_values(by=['lift'], ascending=False)
df_res.head()

antecedents  \
2                                      (50 First Dates)   
3                                    (Along Came Polly)   
3846       (Finding Nemo (Widescreen), The Sixth Sense)   
3843                (The Silence of the Lambs, Shrek 2)   
4195  (The Silence of the Lambs, Lord of the Rings: ...   

                                            consequents  antecedent support  \
2                                    (Along Came Polly)            0.548799   
3                                      (50 First Dates)            0.390201   
3846                (The Silence of the Lambs, Shrek 2)            0.507890   
3843       (Finding Nemo (Widescreen), The Sixth Sense)            0.428324   
4195  (Braveheart, Pirates of the Caribbean: The Cur...            0.492651   

      consequent support   support  confidence      lift  leverage  conviction  
2               0.390201  0.314667    0.573374  1.469433  0.100525    1.429353  
3               0.548799  0.314667    0.806423  1.469433  0.100525    2.330864  
3846            0.428324  0.314176    0.618589  1.444209  0.096634    1.498846  
3843            0.507890  0.314176    0.733500  1.444209  0.096634    1.846563  
4195            0.435541  0.308547    0.626300  1.437980  0.093977    1.510459

In [ ]:
#Finding number of antecedents
df_res["antecedents_len"] = df_res["antecedents"].apply(lambda x: len(x))
df_res.head()

antecedents  \
2                                      (50 First Dates)   
3                                    (Along Came Polly)   
3846       (Finding Nemo (Widescreen), The Sixth Sense)   
3843                (The Silence of the Lambs, Shrek 2)   
4195  (The Silence of the Lambs, Lord of the Rings: ...   

                                            consequents  antecedent support  \
2                                    (Along Came Polly)            0.548799   
3                                      (50 First Dates)            0.390201   
3846                (The Silence of the Lambs, Shrek 2)            0.507890   
3843       (Finding Nemo (Widescreen), The Sixth Sense)            0.428324   
4195  (Braveheart, Pirates of the Caribbean: The Cur...            0.492651   

      consequent support   support  confidence      lift  leverage  \
2               0.390201  0.314667    0.573374  1.469433  0.100525   
3               0.548799  0.314667    0.806423  1.469433  0.100525   
3846            0.428324  0.314176    0.618589  1.444209  0.096634   
3843            0.507890  0.314176    0.733500  1.444209  0.096634   
4195            0.435541  0.308547    0.626300  1.437980  0.093977   

      conviction  antecedents_len  
2       1.429353                1  
3       2.330864                1  
3846    1.498846                2  
3843    1.846563                2  
4195    1.510459                2

In [ ]:
#Finding number of consequents
df_res["consequents_len"] = df_res["consequents"].apply(lambda x: len(x))
df_res.head()

antecedents  \
2                                      (50 First Dates)   
3                                    (Along Came Polly)   
3846       (Finding Nemo (Widescreen), The Sixth Sense)   
3843                (The Silence of the Lambs, Shrek 2)   
4195  (The Silence of the Lambs, Lord of the Rings: ...   

                                            consequents  antecedent support  \
2                                    (Along Came Polly)            0.548799   
3                                      (50 First Dates)            0.390201   
3846                (The Silence of the Lambs, Shrek 2)            0.507890   
3843       (Finding Nemo (Widescreen), The Sixth Sense)            0.428324   
4195  (Braveheart, Pirates of the Caribbean: The Cur...            0.492651   

      consequent support   support  confidence      lift  leverage  \
2               0.390201  0.314667    0.573374  1.469433  0.100525   
3               0.548799  0.314667    0.806423  1.469433  0.100525   
3846            0.428324  0.314176    0.618589  1.444209  0.096634   
3843            0.507890  0.314176    0.733500  1.444209  0.096634   
4195            0.435541  0.308547    0.626300  1.437980  0.093977   

      conviction  antecedents_len  consequents_len  
2       1.429353                1                1  
3       2.330864                1                1  
3846    1.498846                2                2  
3843    1.846563                2                2  
4195    1.510459                2                3

Below code is for demo:
Convert the antecedents generated into a list from a frozen set

In [ ]:
df_res['antecedents'] = df_res['antecedents'].apply(list)

Keep only single antecedents

In [ ]:
df_res=df_res[(df_res['antecedents_len'] == 1)]
df_res

antecedents  \
2           [50 First Dates]   
3         [Along Came Polly]   
686   [When Harry Met Sally]   
687   [Sleepless in Seattle]   
762   [The Matrix: Reloaded]   
...                      ...   
141   [Being John Malkovich]   
1237       [American Beauty]   
92          [Bruce Almighty]   
93         [American Beauty]   
799        [American Beauty]   

                                            consequents  antecedent support  \
2                                    (Along Came Polly)            0.548799   
3                                      (50 First Dates)            0.390201   
686                              (Sleepless in Seattle)            0.419354   
687                              (When Harry Met Sally)            0.498721   
762                                  (X2: X-Men United)            0.464780   
...                                                 ...                 ...   
141   (Pirates of the Caribbean: The Curse of the Bl...            0.411838   
1237  (Pirates of the Caribbean: The Curse of the Bl...            0.642921   
92                                    (American Beauty)            0.591852   
93                                     (Bruce Almighty)            0.642921   
799   (50 First Dates, Pirates of the Caribbean: The...            0.642921   

      consequent support   support  confidence      lift  leverage  \
2               0.390201  0.314667    0.573374  1.469433  0.100525   
3               0.548799  0.314667    0.806423  1.469433  0.100525   
686             0.498721  0.300233    0.715941  1.435554  0.091092   
687             0.419354  0.300233    0.602006  1.435554  0.091092   
762             0.492644  0.323038    0.695035  1.410826  0.094067   
...                  ...       ...         ...       ...       ...   
141             0.787241  0.327285    0.794693  1.009465  0.003069   
1237            0.544239  0.352334    0.548021  1.006949  0.002431   
92              0.642921  0.381352    0.644337  1.002203  0.000838   
93              0.591852  0.381352    0.593156  1.002203  0.000838   
799             0.471092  0.303311    0.471770  1.001438  0.000436   

      conviction  antecedents_len  consequents_len  
2       1.429353                1                1  
3       2.330864                1                1  
686     1.764699                1                1  
687     1.458929                1                1  
762     1.663653                1                1  
...          ...              ...              ...  
141     1.036293                1                1  
1237    1.008367                1                2  
92      1.003982                1                1  
93      1.003204                1                1  
799     1.001283                1                2  

[2241 rows x 11 columns]

In [ ]:
df_res[(df_res['consequents_len'] > 1)]

antecedents                                        consequents  \
3697  [X2: X-Men United]  (Finding Nemo (Widescreen), Pirates of the Car...   
834     [50 First Dates]                          (Bruce Almighty, Shrek 2)   
4046  [X2: X-Men United]  (Pirates of the Caribbean: The Curse of the Bl...   
2526  [X2: X-Men United]  (Lord of the Rings: The Fellowship of the Ring...   
2996           [Shrek 2]  (50 First Dates, Pirates of the Caribbean: The...   
...                  ...                                                ...   
1255   [American Beauty]  (The Last Samurai, Pirates of the Caribbean: T...   
1141   [American Beauty]  (Bruce Almighty, Pirates of the Caribbean: The...   
1165   [American Beauty]               (Finding Nemo (Widescreen), Shrek 2)   
1237   [American Beauty]  (Pirates of the Caribbean: The Curse of the Bl...   
799    [American Beauty]  (50 First Dates, Pirates of the Caribbean: The...   

      antecedent support  consequent support   support  confidence      lift  \
3697            0.492644            0.458247  0.311775    0.632860  1.381046   
834             0.548799            0.433910  0.324948    0.592107  1.364586   
4046            0.492644            0.472147  0.316035    0.641508  1.358705   
2526            0.492644            0.453979  0.302890    0.614826  1.354304   
2996            0.627781            0.363427  0.307585    0.489956  1.348156   
...                  ...                 ...       ...         ...       ...   
1255            0.642921            0.493157  0.325411    0.506145  1.026336   
1141            0.642921            0.509422  0.332984    0.517924  1.016690   
1165            0.642921            0.494461  0.321428    0.499950  1.011102   
1237            0.642921            0.544239  0.352334    0.548021  1.006949   
799             0.642921            0.471092  0.303311    0.471770  1.001438   

      leverage  conviction  antecedents_len  consequents_len  
3697  0.086022    1.475603                1                3  
834   0.086819    1.387840                1                2  
4046  0.083435    1.472427                1                3  
2526  0.079240    1.417595                1                2  
2996  0.079433    1.248075                1                3  
...        ...         ...              ...              ...  
1255  0.008350    1.026299                1                2  
1141  0.005466    1.017636                1                2  
1165  0.003529    1.010978                1                2  
1237  0.002431    1.008367                1                2  
799   0.000436    1.001283                1                2  

[1447 rows x 11 columns]

Split the consequents into multiple rows

In [ ]:
df_res=df_res.explode('consequents')
df_res.tail()

antecedents                                        consequents  \
1237  [American Beauty]                                            Shrek 2   
92     [Bruce Almighty]                                    American Beauty   
93    [American Beauty]                                     Bruce Almighty   
799   [American Beauty]                                     50 First Dates   
799   [American Beauty]  Pirates of the Caribbean: The Curse of the Bla...   

      antecedent support  consequent support   support  confidence      lift  \
1237            0.642921            0.544239  0.352334    0.548021  1.006949   
92              0.591852            0.642921  0.381352    0.644337  1.002203   
93              0.642921            0.591852  0.381352    0.593156  1.002203   
799             0.642921            0.471092  0.303311    0.471770  1.001438   
799             0.642921            0.471092  0.303311    0.471770  1.001438   

      leverage  conviction  antecedents_len  consequents_len  
1237  0.002431    1.008367                1                2  
92    0.000838    1.003982                1                1  
93    0.000838    1.003204                1                1  
799   0.000436    1.001283                1                2  
799   0.000436    1.001283                1                2